src/btree/bt_delete.c

```diff
@@ -315,13 +330,15 @@ __wt_delete_page_instantiate(WT_SESSION_IMPL *session, WT_REF *ref)
         WT_STAT_CONN_DATA_INCR(session, cache_read_deleted_prepared);
 
     /*
-     * Give the page a modify structure. If the tree is already dirty and so will be written, mark
-     * the page dirty. (We want to free the deleted pages, but if the handle is read-only or if the
-     * application never modifies the tree, we're not able to do so.)
+     * Give the page a modify structure and mark the page dirty if the tree isn't read-only. If the
+     * tree can be written, the page must be marked dirty because if we discard it as a clean page,
+     * then read it back into memory, we'll have the same problem as when instantiating deleted
+     * pages in read-only trees, re-reading the page won't re-create the fast-truncate information.
+     * If the tree cannot be written (checked in page-modify-set), we don't dirty page and we don't
+     * free the ft_info.del field below.
      */
     WT_RET(__wt_page_modify_init(session, page));
-    if (btree->modified)
-        __wt_page_modify_set(session, page);
+    __wt_page_modify_set(session, page);
```

Comment:
And this the problem that caused the original reversion, it causes checkpoints to mark the tree dirty.

Reply:
I think we can get past this problem with:
```
if (FLD_ISSET(session->lock_flags, WT_SESSION_LOCKED_CHECKPOINT))
```
but  it's not correct.

Reply:
(this is obsolete given your patch today that fixes WT_READ_CACHE) (that is, [94cdc53](https://github.com/wiredtiger/wiredtiger/pull/7909/commits/94cdc53ddd39ad63958e32ba945d4e5255bd2d95))

Reply:
I'm chiming in here 20 days late, because I just came across a failure related to this. I found in WT-9323 that queueing the page for eviction soon triggers a data consistency problem.

I believe the change you have here will mask that - in that it seemed to only be applicable for clean pages. In the past we've oscillated about when it's OK for this code to mark a page dirty. I can see that this is safe now, since `wt_page_modify_set` doesn't do anything if the tree is read-only.

I believe that leads to a bug at the moment. Where if that page is left "clean" and then evicted it will be put into WT_REF_DISK and when read back in, the tombstones won't be created (i.e: the truncated data will come back to life). In a read-only tree that content can't be written back to disk again, but queries would see unexpected results.

It's possibly a bit of a big question for this PR, but I'm becoming inclined to never mark these pages as dirty. I don't think of them as dirty, and I believe the reason we used to do that is that they are more like dirty pages in terms of structure (having lots of small update structures associated with them) and memory allocator fragmentation. I believe that issue was resolved when we added in the eviction mode that tracks updates separately to dirty pages (since they are more common with durable history).

i.e: as long as the memory allocated for tombstones is accounted for in the updates tracking, I think we should leave these pages marked as clean. Which is their "real" state - since there haven't been any changes that need to be reconciled back to disk.

Reply:
There are 3 parts to this change that need to be considered:
* the code in deleted-page-instantiation that marks pages dirty,
* the code in page-read that SKIPS the deleted-page-instantiation call in the case of initially created, empty tree,
* the code in page-can-evict that check for fast-truncate information (which checkpoint asserts when it's discarding trees from memory).

Reply:
> I believe that leads to a bug at the moment. Where if that page is left "clean" and then evicted it will be put into WT_REF_DISK and when read back in, the tombstones won't be created (i.e: the truncated data will come back to life). In a read-only tree that content can't be written back to disk again, but queries would see unexpected results.

Yes, that is a bug we chased down some time ago, I think.

The way this works is when we retrieve the read address in the page-read function (a call to `__wt_ref_addr_copy()`), we will restore the fast-deleted information to the WT_REF based on the internal page's cell, and so the tombstones will be re-created correctly.

As you note, the content cannot be written back to disk; we went through a few attempts to figure this out, and eventually decided that because we already have to crack the internal cell in this case, restoring the fast-truncate information during each subsequent read was the simplest approach.

Reply:
I'm probably missing something obvious, but I think we have to mark the pages dirty if the fast-truncate is not yet committed?

Reply:
Currently we need to mark the page dirty after instantiation because, although the instantiation itself is an identity transformation and can be repeated as many times as we want, the _truncation_ is not and needs to not get lost. When the page is in WT_REF_DELETED state, the truncation is a piece of state that belongs to the parent page. Bugs can cause it to get lost there (e.g. the last set of fixes to use leave_dirty on affected internal pages was about that) but obviously it's not supposed to be.

However, instantiation moves that state from the parent internal page to the child leaf page. After instantiation the leaf page is an ordinary leaf page (it is not _quite_ fully ordinary until it gets reconciled again in case we have to write the parent page out first, but otherwise it's ordinary) and consequently the dirty-state associated with the truncation moves from the parent to the child. If we don't mark the child dirty at that point and it gets discarded we'll lose the truncation.

Keeping the truncation state in the parent after instantiation is hard and messy; we did it for readonly trees at first but got rid of that as it was making the ref->ft_info union accesses too complicated to get right. Furthermore, if the truncation isn't committed we need to use the pointer there to hold the update list so we don't have it available to hang onto the truncation information itself. (Though that could be fixed by doing more allocations or moving the update list inside page_del.) But also there's another problem, which is that the page skipping has to stop paying attention to the delete info as soon as the child becomes dirty for real, and at that point we also have to avoid writing out a deleted-address cell when reconciling the parent page so the new changes don't accidentally get subsumed into the truncation. Adding code to clear the delete info when we mark pages dirty is possible I guess but it at least seems likely to create locking problems and races.